pkg/controller/populators/forklift-populator.go

```diff
@@ -0,0 +1,593 @@
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	"kubevirt.io/containerized-data-importer-api/pkg/apis/forklift/v1beta1"
+	cc "kubevirt.io/containerized-data-importer/pkg/controller/common"
+	featuregates "kubevirt.io/containerized-data-importer/pkg/feature-gates"
+)
+
+const (
+	forkliftPopulatorName  = "forklift-populator-controller"
+	populatorContainerName = "populate"
+	populatorPodPrefix     = "populate"
+	populatorPodVolumeName = "target"
+	mountPath              = "/mnt/"
+	devicePath             = "/dev/block"
+)
+
+const apiGroup = "forklift.kubevirt.io"
+
+var (
+	supportedPopulators = map[string]client.Object{
+		"OvirtVolumePopulator":     &v1beta1.OvirtVolumePopulator{},
+		"OpenstackVolumePopulator": &v1beta1.OpenstackVolumePopulator{},
+	}
+)
+
+var errCrNotFound = errors.New("populator CR not found")
+
+// ForkliftPopulatorReconciler members
+type ForkliftPopulatorReconciler struct {
+	ReconcilerBase
+	importerImage       string
+	ovirtPopulatorImage string
+}
+
+// NewForkliftPopulator creates a new instance of the forklift controller.
+func NewForkliftPopulator(
+	ctx context.Context,
+	mgr manager.Manager,
+	log logr.Logger,
+	importerImage string,
+	ovirtPopulatorImage string,
+	installerLabels map[string]string,
+) (controller.Controller, error) {
+	client := mgr.GetClient()
+	reconciler := &ForkliftPopulatorReconciler{
+		ReconcilerBase: ReconcilerBase{
+			client:          client,
+			scheme:          mgr.GetScheme(),
+			log:             log.WithName(forkliftPopulatorName),
+			recorder:        mgr.GetEventRecorderFor(forkliftPopulatorName),
+			featureGates:    featuregates.NewFeatureGates(client),
+			installerLabels: installerLabels,
+		},
+		importerImage:       importerImage,
+		ovirtPopulatorImage: ovirtPopulatorImage,
+	}
+
+	forkliftPopulatorController, err := controller.New(forkliftPopulatorName, mgr, controller.Options{
+		Reconciler: reconciler,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	for kind, sourceType := range supportedPopulators {
+		if err := addWatchers(mgr, forkliftPopulatorController, log, kind, sourceType); err != nil {
+			return nil, err
+		}
+	}
+
+	return forkliftPopulatorController, nil
+}
+
+func addWatchers(mgr manager.Manager, c controller.Controller, log logr.Logger, sourceKind string, sourceType client.Object) error {
+	// Setup watches
+	if err := c.Watch(source.Kind(mgr.GetCache(), &corev1.PersistentVolumeClaim{}), handler.EnqueueRequestsFromMapFunc(
+		func(_ context.Context, obj client.Object) []reconcile.Request {
+			pvc := obj.(*corev1.PersistentVolumeClaim)
+			if isPVCForkliftKind(pvc) {
+				pvcKey := types.NamespacedName{Namespace: pvc.Namespace, Name: pvc.Name}
+				return []reconcile.Request{{NamespacedName: pvcKey}}
+			}
+
+			// TODO(benny) check dataSourceRef PVC prime
+			if isPVCForkliftKind(pvc) {
```

Comment:
How is this ever getting called. If the condition is true, the above if will return out of the calling function.

Reply:
yeah it should watch the prime pvc

Reply:
But if you look at the code, the if statement above is identical in the check. And it returns out of the function. So if it matches, the second if is never called, and if it doesn't match, the second if will also never match.

Reply:
yes, I meant that it's going to be changed to watch the prime pvc like the other populators do